processing/src/main/java/org/apache/druid/query/aggregation/SingleValueBufferAggregator.java

```diff
@@ -0,0 +1,155 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.query.aggregation;
+
+import org.apache.druid.common.config.NullHandling;
+import org.apache.druid.error.InvalidInput;
+import org.apache.druid.segment.ColumnValueSelector;
+import org.apache.druid.segment.DimensionHandlerUtils;
+import org.apache.druid.segment.column.ColumnType;
+import org.apache.druid.segment.column.ValueType;
+
+import javax.annotation.Nullable;
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+
+/**
+ *
+ */
+public class SingleValueBufferAggregator implements BufferAggregator
+{
+  final ColumnValueSelector selector;
+
+  final ColumnType columnType;
+
+  private int stringByteArrayLength = 0;
+
+  private boolean isAggregateInvoked = false;
+
+  SingleValueBufferAggregator(ColumnValueSelector selector, ColumnType columnType)
+  {
+    this.selector = selector;
+    this.columnType = columnType;
+  }
+
+  @Override
+  public void init(ByteBuffer buf, int position)
+  {
+    buf.put(position, NullHandling.IS_NULL_BYTE);
+  }
+
+  @Override
+  public void aggregate(ByteBuffer buf, int position)
+  {
+    if (isAggregateInvoked) {
+      throw InvalidInput.exception("Single Value Aggregator would not be applied to more than one row");
+    }
+    boolean isNotNull;
+    if (columnType.is(ValueType.STRING)) {
+      isNotNull = (selector.getObject() != null);
+    } else {
+      isNotNull = !selector.isNull();
+    }
+    if (isNotNull) {
+      if (buf.get(position) == NullHandling.IS_NULL_BYTE) {
+        buf.put(position, NullHandling.IS_NOT_NULL_BYTE);
+      }
+      updatevalue(buf, position + Byte.BYTES);
+    }
```

Comment:
Part of the potentially attractive part about pushing the `ColumnType` in here is that we can also use the `NullableTypeStrategy` to implement `aggregate` and `get` methods.

I pulled your branch and tried this and it runs into some problems with floats ending up as doubles, though I think this is actually a problem with either the underlying expression selector (expressions don't support floats so I think this is likely the culprit) because `testSingleValueFloatAgg` fails, so we need a helper function to get the object from the underlying selector to ensure that the thing we get from the selector matches the `columnType`.

```
  @Override
  public void aggregate(ByteBuffer buf, int position)
  {
    if (isAggregateInvoked) {
      throw InvalidInput.exception("Single Value Aggregator would not be applied to more than one row");
    }

    int written = typeStrategy.write(
        buf,
        position,
        getSelectorObject(),
        SingleValueAggregatoractory.DEFAULT_MAX_STRING_SIZE
    );
    if (written < 0) {
      throw InvalidInput.exception("Single Value Aggregator value too big for buffer");
    }

    isAggregateInvoked = true;
  }
```

(invalid input might not be the actual right exception here for exceeding size limit, since its only a single row we might want to figure out how to pass this in, or make a much larger limit since the user isn't calling this function directly..)

```
  @Nullable
  @Override
  public Object get(ByteBuffer buf, int position)
  {
    return typeStrategy.read(buf, position);
  }
```


```
  @Nullable
  private Object getSelectorObject()
  {
    if (selector.isNull()) {
      return null;
    }
    switch (columnType.getType()) {
      case LONG:
        return selector.getLong();
      case FLOAT:
        return selector.getFloat();
      case DOUBLE:
        return selector.getDouble();
      default:
        return selector.getObject();
    }
  }
``` 

The `getLong`/`getFloat`/`getDouble` methods could also be updated, there are some static methods in `TypeStrategies` which can help with this `isNullableNull`, `readNotNullNullableLong`, etc.

Reply:
would look more cleaner the way you pointed. will update it.